ludwig/models/llm.py

```diff
@@ -418,23 +489,117 @@ def get_args(self):
+            return input_ids, torch.stack(attention_masks)
+
+        merged_input_and_targets = []
+        lengths = []
+
+        pad_tensor = torch.tensor([self.tokenizer.pad_token_id]).to(target_ids[0].device)
+
+        # Merge input_ids and target_ids by concatenating them together.
+        # We remove the left padding from both input_ids and target_ids before concatenating them.
+        for input_id_sample, target_id_sample in zip(input_ids, target_ids):
+            input_id_sample_no_padding = self._remove_left_padding(input_id_sample)[0]
+            target_id_sample_no_padding = torch.cat(
+                (self._remove_left_padding(target_id_sample)[0], pad_tensor), dim=-1
+            )
+
+            merged_sample_ids = torch.cat((input_id_sample_no_padding, target_id_sample_no_padding), dim=-1)
+
+            merged_input_and_targets.append(merged_sample_ids)
+            lengths.append(merged_sample_ids.shape[0])
+
+        # Since we remove the left padding from the target_ids, the merged input_ids and target_ids
+        # may not have the same lengths. We need to align them to the same length by adding left padding
+        # and generate an attention mask for just the part of the input that is not padding.
+        max_length = max(lengths)
+        attention_masks = []
+        for i, merged_sample_ids in enumerate(merged_input_and_targets):
+            merged_input_and_targets[i] = self._add_left_padding(merged_sample_ids, max_length)
+            attention_masks.append(self._create_attention_mask(merged_input_and_targets[i]))
+
+        return torch.stack(merged_input_and_targets), torch.stack(attention_masks)
+
     def _remove_left_padding(self, input_ids_sample: torch.Tensor):
-        bos_idxs = torch.where(input_ids_sample == self.tokenizer.bos_token_id)[0]  # all BOS token locations
+        """Removes left padding from the input_ids tensor."""
+        # Remove all PAD tokens
+        pad_idxs = torch.where(input_ids_sample == self.tokenizer.pad_token_id)[0]  # all PAD token locations
+        if len(pad_idxs) != 0:
+            pad_idx = pad_idxs[-1]  # get last PAD token location
+        else:
+            pad_idx = 0
+        input_ids_sample_no_padding = input_ids_sample[pad_idx + 1 :]
+
+        # Start from the first BOS token
+        bos_idxs = torch.where(input_ids_sample_no_padding == self.tokenizer.bos_token_id)[0]  # all BOS token locations
         if len(bos_idxs) != 0:
             bos_idx = bos_idxs[0]  # get first BOS token location
         else:
             bos_idx = 0
 
-        input_ids_sample_no_padding = input_ids_sample[bos_idx:].unsqueeze(0)
-        return input_ids_sample_no_padding
+        input_ids_sample_no_bos = input_ids_sample_no_padding[bos_idx:].unsqueeze(0)
+        return input_ids_sample_no_bos
+
+    def _add_left_padding(self, input_ids, max_length, pad_value=0):
+        """Adds left padding to the input_ids tensor."""
+        if not pad_value:
```

Comment:
couldn't we always do the `else` block?

```
padding = torch.tensor(
  [pad_value] * (max_length - input_ids.shape[0]), dtype=torch.int32, device=input_ids.device
)
```

since pad_value being 0 would be equivalent?